### DISCUSSION TITLE
Some confusion about signature_algorithms

### DISCUSSION THREAD

**Discussion by DreamDLYY** [openssl/openssl#26110] [Q&A] [answered]
Some confusion about signature_algorithms

Thank you for reading these questions.
I was recently learning about the TLS1.2 protocol flow and communication, and I captured some TLS handshake packets to read and analyze, but was confused about the signature_algorithms.
1. In the client hello extension, signature_algorithms is described in the certificates section of rfc5246 7.4.2 
as follows:

‘’‘
   If the client provided a "signature_algorithms" extension, then all
   certificates provided by the server MUST be signed by a
   hash/signature algorithm pair that appears in that extension.  Note
   that this implies that a certificate containing a key for one
   signature algorithm MAY be signed using a different signature
   algorithm (for instance, an RSA key signed with a DSA key).  This is
   a departure from TLS 1.1, which required that the algorithms be the
   same.  Note that this also implies that the DH_DSS, DH_RSA,
   ECDH_ECDSA, and ECDH_RSA key exchange algorithms do not restrict the
   algorithm used to sign the certificate.  Fixed DH certificates MAY be
   signed with any hash/signature algorithm pair appearing in the
   extension.  The names DH_DSS, DH_RSA, ECDH_ECDSA, and ECDH_RSA are
   historical.
’‘’

But shouldn't the algorithm used to sign a certificate be determined at the time the certificate is issued? The certificate should be signed by the trusted certificate above the current certificate.Why is signing possible here?

2. When using a key exchange algorithm like ECHDE, there is a server key exchange TLS structure, and in its Server Params I found a parameter called  Signature Algorithm:

TLSv1.2 Record Layer: Handshake Protocol: Server Key Exchange
    Content Type: Handshake (22)
    Version: TLS 1.2 (0x0303)
    Length: 333
    Handshake Protocol: Server Key Exchange
        Handshake Type: Server Key Exchange (12)
        Length: 329
        EC Diffie-Hellman Server Params
            Curve Type: named_curve (0x03)
            Named Curve: secp256r1 (0x0017)
            Pubkey Length: 65
            Pubkey: 04081a569471bec78fee8c7bd83a89067ce6e2bb5ff3627978df5e554f5b012de255dcd6…
            Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                Signature Hash Algorithm Hash: SHA512 (6)
                Signature Hash Algorithm Signature: RSA (1)
            Signature Length: 256
            Signature: 8b5175ac06ecbde95628be21ff08ce904b14685a0f03463a75e68040e1cdd67749ae9793…

I'm not sure what this parameter does. In my understanding, the contents of the Server Params should be used for key exchange, and the signature should be implemented by the signing algorithm of the certificate.

I hope who can answer my doubts, thanks a lot.

**Comment by mattcaswell** [comment, marked as answer]
> But shouldn't the algorithm used to sign a certificate be determined at the time the certificate is issued?

It is.

>The certificate should be signed by the trusted certificate above the current certificate

Correct

> Why is signing possible here?

This isn't about doing a sign operation. This is about the client declaring what signature algorithms it is prepared to process, and the server selecting a certificate to send back that meets those criteria. In practice a server is only configured with 1 or perhaps 2 certificates anyway - so they will usually respond with what they've got anyway regardless of what the RFC says.

The RFC goes on to say:

>    If the server has multiple certificates, it chooses one of them based
   on the above-mentioned criteria (in addition to other criteria, such
   as transport layer endpoint, local configuration and preferences,
   etc.).  If the server has a single certificate, it SHOULD attempt to
   validate that it meets these criteria

IIRC OpenSSL does not validate this (its a SHOULD not a MUST)

>I'm not sure what this parameter does. In my understanding, the contents of the Server Params should be used for key exchange, and the signature should be implemented by the signing algorithm of the certificate.

Key exchange by itself provides no authentication. Here the server is signing the ServerKeyExchange message using the private key associated with its certificate in order to provide that authentication. This is actually a primary use case for the sig_algs extension. It will select one of the sig algs from the extension that the client sent in its client hello that is consistent with the key associated with its cert, and then create a signature according to that sig alg.

**Reply by DreamDLYY** [reply]
Thank you very much for your reply, maybe it is the reason I stated, there are actually two questions here, the first question has been perfectly solved by you, and the second question is not related to the "signature algorithms extension" of "client hello".
As you said, “Key exchange by itself provides no authentication.Here the server is signing the ServerKeyExchange message using the  private key associated with its certificate in order to provide that authentication. "My understanding is the same as yours.
But in the packet data that I captured, The "server key exchange" packet contains a two-byte "Signature Algorithm" that is "rsa_pkcs1_sha512 (0x0601)" under the "EC Diffie-Hellman Server Params" parameter.
I don't understand what the purpose of the "Signature Algorithm" in this place is.
Since for communication, the signature algorithm used should be the signature algorithm in the certificate, what is the effect of the existence of this signature algorithm structure in this structure?
 I've indented the data in this packet to make it easier to read.

```
TLSv1.2 Record Layer: Handshake Protocol: Server Key Exchange
    Content Type: Handshake (22)
    Version: TLS 1.2 (0x0303)
    Length: 333
        Handshake Protocol: Server Key Exchange
        Handshake Type: Server Key Exchange (12)
        Length: 329
        EC Diffie-Hellman Server Params
            Curve Type: named_curve (0x03)
            Named Curve: secp256r1 (0x0017)
            Pubkey Length: 65
            Pubkey: 04081a569471bec78fee8c7bd83a89067ce6e2bb5ff3627978df5e554f5b012de255dcd6…
            Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                Signature Hash Algorithm Hash: SHA512 (6)
                Signature Hash Algorithm Signature: RSA (1)
            Signature Length: 256
            Signature: 8b5175ac06ecbde95628be21ff08ce904b14685a0f03463a75e68040e1cdd67749ae9793…
```

As far as I know, the "signature length" and "signature" of the fourth and fifth last lines are determined by the signature algorithm in the certificate. What confuses me is this：

            Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                Signature Hash Algorithm Hash: SHA512 (6)
                Signature Hash Algorithm Signature: RSA (1)

Another: THIS text is translated by the translation software,  I tried to express clearly, but there may be grammar is not smooth or inappropriate places, if there is anything not clear, please ask me, forgive me. ;)

**Reply by DreamDLYY** [reply]
P.S. I found a similar structure in Certificate Verify section:

```
TLSv1.2 Record Layer: Handshake Protocol: Certificate Verify
    Content Type: Handshake (22)
    Version: TLS 1.2 (0x0303)
    Length: 264
    Handshake Protocol: Certificate Verify
        Handshake Type: Certificate Verify (15)
        Length: 260
        Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
            Signature Hash Algorithm Hash: SHA256 (4)
            Signature Hash Algorithm Signature: RSA (1)
        Signature length: 256
        Signature: 3e8776405c95b6f1737a8c52748e522350b99bdf5a712609050ee73ac89880a658b9b89e…
```

I don't see the point of this part either:

        Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
            Signature Hash Algorithm Hash: SHA256 (4)
            Signature Hash Algorithm Signature: RSA (1)

**Reply by mattcaswell** [reply]
> Since for communication, the signature algorithm used should be the signature algorithm in the certificate, what is the effect of the existence of this signature algorithm structure in this structure?

The certificate will specify the underlying algorithm. In this case RSA. But there are more components than just the core underlying algorithm. Most notably the hash function - that is not determined by the certificate. It's also possible for the sig alg to specify PSS padding instead of PKCS#1 1.5 padding.

**Reply by DreamDLYY** [reply]
So the sig alg in this Server Key Exchange is used for padding? Is sig alg in the Certificate Verify example also used as padding?

**Reply by mattcaswell** [reply]
The signature itself must always be padded using some padding scheme. That's just the way RSA signatures work. The choices for an RSA signature are PKCS#1 v1.5 padding or PSS padding. This applies to both the signature in the Server Key Exchange message, or the one in the Certificate Verify message.

**Reply by DreamDLYY** [reply]
thank you a lot！ I see the problem here, the hash algorithm here is used to calculate the hash value of the subsequent signature, right? I mistaking the length of the signature in the latter part for the length after the hash calculation, which should be the length after the RSA signature.

**Reply by mattcaswell** [reply]
> the hash algorithm here is used to calculate the hash value of the subsequent signature, right?

To be clear, the message to be signed is first hashed by the hash algorithm. The resulting hash value is what is actually signed (not the message itself).

**Reply by DreamDLYY** [reply]
I get, Thank you~
